Visual/visual.py

```diff
@@ -0,0 +1,664 @@
+             >>> n=3
+             >>> x = Variable(n)
+             >>> P = cvxopt.matrix([13, 12, -2,12, 17, 6,-2, 6, 12], (n, n))
+             >>> objective = Minimize(0.5 * quad_form(x, P))
+             >>> v = Visual(objective)
+             >>> v.root.curvature == 'CONVEX'
+             True
+             >>> v.root.sign == 'POSITIVE'
+             True
+             >>> v.root.sons[0].sons[0].curvature == 'CONSTANT'
+             True
+             >>> v.root.sons[0].sons[0].sign == 'POSITIVE'
+             True
+             >>> v.root.sons[0].sons[1].curvature == 'CONVEX'
+             True
+             >>> v.root.sons[0].sons[1].sign == 'POSITIVE'
+             True
+             """
+        # If the value of the node is an expression
+        if node.expr is not None and node.expr is not False and node.flag == 0:
+            # We will go through the list and look for the value of the current node
+            for arg in self.curvature_sign_list:
+                node_exp = node.expr
+                if str(arg[0])[0] == '-':
+                    node_exp = '-' + node.expr
+                node_exp = node_exp.replace(' + -', ' - ').replace('\n', ' ').replace(' ', '')
+                arg_exp = str(arg[0]).replace(' + -', ' - ').replace('\n', ' ').replace(' ', '')
+                # If we found the value then we will insert the corresponding values into the node
+                if node_exp.__eq__(arg_exp):
+                    node.curvature = arg[1]
+                    # To present things more clearly
+                    if str(arg[2]).__eq__("NONNEGATIVE"):
+                        arg[2] = "POSITIVE"
+                    if str(arg[2]).__eq__("NONPOSITIVE"):
+                        arg[3] = "NEGATIVE"
+                    node.sign = arg[2]
+                    break
+        # We will go over the children of the node and do the same
+        for child in node.sons:
+            self.curvature_sign_node(child)
+
+    def plot_function(self, expr, xmin, xmax, num_points=1000):
+        func, xmin, xmax = self.string_to_plot(self.expr)
+        x_vals = np.linspace(xmin, xmax, num_points)
+        y_vals = func(x_vals)
+        plt.plot(x_vals, y_vals)
+        plt.xlabel('x')
+        plt.ylabel('y')
+        plt.title('Graph of Function')
+        plt.show()
+
+    def string_to_plot(self, s):
+        # Extract the variable name and expression from the string
+        # Replace power(var, n) with var^n
+        expr = re.sub(r'power\((\w+),\s*([\d\.]+)\)', r'\1**\2', s)
+
+        # Replace '@' with '*'
+        expr = expr.replace('@', '*')
+
+        expr = expr.replace(self.variables[0], 'x')
+
+        # Replace the variable name in the expression with 'x'
+
+        # Replace '^' with '**'
+
+        # Construct the lambda function and return the input for plot_function
+        func = lambda x: eval(expr)
+        return (func, -10, 10)
+
+    def draw_graph(self):
```

Comment:
It should be possible here to change the x limits, so it's not always -10 to 10.

Reply:
we change it, now you can change the x limits when -10 to 10 is the default